Refine minimum string length after startsWith and endsWith

panacekcz opened the pull request:
This is a part of the Bounded-size strings GSoC project, made in this fork to start a review process by  @kelloggm. A PR with this feature will be opened in the main repository after typetools#1352 is merged.

With these changes, the minimum length of the receiver string in the then-branch is refined to the minimum length known for the argument (based on a stored annotation or the length of a string literal).

The purpose is to support code like:
```java
if(str.startsWith("x")){
 String after = str.substring(1);
}
```
This pattern appears often in [plume-lib](https://github.com/typetools/plume-lib) and causes warnings from the index checker with support for strings (typetools#1413). With the added refinement, the checker will see that the string is at least as long as the prefix or suffix and so the index is valid for the string.

Unlike outlined in kelloggm#56, this PR does not use annotations, but adds special handling of `String.startsWith(String)` and `String.endWith(String)` to `ValueTransfer`.

panacekcz committed:
Add tests for startsWith and endsWith length refinement

panacekcz committed:
Fix test

panacekcz committed:
Refine minimum length after startsWith and endsWith

panacekcz committed:
Move methods to ValueMethodIdentifier

kelloggm commented on framework/src/org/checkerframework/common/value/ValueTransfer.java:
```diff
@@ -400,7 +402,7 @@ private void refineStringAtLengthInvocation(
         MethodAccessNode methodAccessNode = stringLengthNode.getTarget();
 
-        if (atypefactory.isStringLengthMethod(methodAccessNode.getMethod())) {
+        if (atypefactory.methods.isStringLengthMethod(methodAccessNode.getMethod())) {
```
I'm very uncomfortable with reading through a field of another object. We generally prefer getters, though in this case the object is `final` so I suppose it might be okay. I think I'd actually prefer to just initialize another `ValueMethodIdentifier` here and use that one's fields, even though there will be a performance hit. Even better would be to make `ValueMethodIdentifier`'s methods be static, so that you don't have to initialize one at all, but I know that would be problematic because you need access to a `processingEnvironment`. So the choice is either to have some dangerous looking code (what you currently have), to create a somewhat redundant getter, initialize another VMI, or to call `TreeUtils.getMethod` every time that one of the VMI's methods is called. I think I prefer the second or third option.

panacekcz replied:
Ok, I don't see any problem with making a getter.

kelloggm commented on framework/src/org/checkerframework/common/value/ValueMethodIdentifier.java:
```diff
@@ -0,0 +1,41 @@
+     * Determines whether a method is the {@code String.startsWith(String)} or {@code
+     * String.endsWith(String)} method.
+     */
+    public boolean isStartsOrEndsWithMethod(ExecutableElement method) {
```
Probably this class should have two methods, and the client should be responsible for the or.

kelloggm reviewed (changes requested):
Make sure that you justify why you can't use the method suggested in kelloggm#56 somewhere in the code, and also make a comment on the issue explaining why that method is flawed (like we discussed the other week).

panacekcz committed:
Access a field through a getter

panacekcz committed:
Split startsWith and endsWith check

panacekcz committed:
Explain why polymorphic qualifier is not used

panacekcz commented:
@kelloggm, I have addressed the review comments.

panacekcz commented:
Does not work in this method:
```java
void m(String str) {
	int i = 0;
	while (str.endsWith("[]")) {
		@MinLen(2) String str2 = str;
		i++;
	}
}
```
```
error: [assignment.type.incompatible] incompatible types in assignment.
			@MinLen(2) String str2 = str;
			                         ^
  found   : @UnknownVal String
  required: @ArrayLenRange(from=2, to=2147483647) String
``` 
Removing the increment or using `if` instead of `while` makes it work.

mernst committed:
Add Shinya Yoshida to list of contributors [ci skip]

panacekcz committed:
Apply @ArrayLen and @ArrayLenRange to strings (#1352)

(This is a first part of the Bounded-size strings GSoC project.)
Summary of changes:
  - Meaning of @ArrayLen and @ArrayLenRange is extended to string values of the specified lengths.
  - @StringVal is a subtype of @ArrayLen of the corresponding lengths.
  - @StringVal with too many values reduces to @ArrayLen or @ArrayLenRange.
  - For the result of String.length(), @IntVal or @IntRange is inferred from the string annotations on the receiver.
  - The result of String.length() is used to refine the length annotation of the receiver.
  - @StringVal, @ArrayLen and @ArrayLenRange are added to return types of methods in statically-executable.astub.
  - Value checker subtyping diagram is updated with a new part showing subtyping of @StringVal, @ArrayLen and @ArrayLenRange.

panacekcz committed:
Merge branch 'bounded-strings-dev' into bounded-strings-startsendswith

Conflicts:
	framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

panacekcz committed:
Merge branch 'master' into bounded-strings-startsendswith

Conflicts:
	framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
	framework/src/org/checkerframework/common/value/ValueTransfer.java

mernst committed:
Change misleading debug output [ci skip]

mernst committed:
Update commands for removing comments [ci skip]

mernst committed:
Improve cross-references

mernst committed:
Document @UnknownInterned in the manual [ci skip]

bitterfox committed:
Fix #753: Flow Expression parser: parse method arguments containing dots (#1396)

Also fixes #1406.

cmackie committed:
Integrate the Value Checker into the Signedness Checker

Also adds new introduction rules for identifiers; needs to be extended to other expressions.

panacekcz committed:
Merge branch 'master' into bounded-strings-startsendswith

panacekcz commented:
Reopened as typetools#1432 in the main repository.
